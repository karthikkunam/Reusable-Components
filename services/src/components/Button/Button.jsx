import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';

function Button(props) {
  const { variant, className, onClick, disabled, outline, size, children, type } = props;
  const finalType = `btn-${outline ? 'outline-' : ''}${variant}`;
  const finalSize = `btn-${size}`;

  return (
    <button
      type={type}
      className={classNames('btn', finalType, { disabled }, { [finalSize]: size }, { ...className })}
      onClick={!disabled && onClick}>
      {children}
    </button>
  );
}

Button.displayName = 'Button';
Button.defaultProps = {
  variant: 'primary',
  disabled: false,
  outline: false,
  size: undefined,
  className: '',
  onClick: () => {},
  children: undefined,
};

Button.propTypes = {
  variant: PropTypes.oneOf(['primary', 'secondary', 'success', 'danger', 'warning', 'info', 'light', 'dark', 'link']),
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  outline: PropTypes.bool,
  size: PropTypes.oneOf(['lg', 'sm']),
  className: PropTypes.string,
  children: PropTypes.node,
  type: PropTypes.string
};

export default Button;
