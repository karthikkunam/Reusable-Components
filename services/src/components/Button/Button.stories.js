import React from 'react';
import { action } from '@storybook/addon-actions';
import { Button, ButtonSwitch } from '.';
import { withKnobs } from '@storybook/addon-knobs';
import { select, boolean } from '@storybook/addon-knobs';

const wrapperStyle = {
  display: 'flex',
  flexWrap: 'wrap',
  justifyContent: 'space-between',
};
export default {
  title: 'Button',
  component: Button,
  decorators: [storyFn => <div style={wrapperStyle}>{storyFn()}</div>, withKnobs],
};

export const Variant = () => (
  <>
    <Button onClick={action('clicked')} variant="primary">
      Primary
    </Button>
    <br />
    <Button onClick={action('clicked')} variant="secondary">
      Secondary
    </Button>
    <br />
    <Button onClick={action('clicked')} variant="success">
      Success
    </Button>
    <br />
    <Button onClick={action('clicked')} variant="danger">
      danger
    </Button>
    <br />
    <Button onClick={action('clicked')} variant="warning">
      warning
    </Button>
    <br />
    <Button onClick={action('clicked')} variant="info">
      info
    </Button>
    <br />
    <Button onClick={action('clicked')} variant="light">
      light
    </Button>
    <br />
    <Button onClick={action('clicked')} variant="dark">
      dark
    </Button>
    <br />
    <Button onClick={action('clicked')} variant="link">
      link
    </Button>
    <br />
  </>
);

export const Outline = () => (
  <>
    <Button onClick={action('clicked')} variant="primary" outline>
      Primary
    </Button>
    <Button onClick={action('clicked')} variant="secondary" outline>
      Secondary
    </Button>
    <Button onClick={action('clicked')} variant="success" outline>
      Success
    </Button>
    <Button onClick={action('clicked')} variant="danger" outline>
      danger
    </Button>
    <Button onClick={action('clicked')} variant="warning" outline>
      warning
    </Button>
    <Button onClick={action('clicked')} variant="info" outline>
      info
    </Button>
    <Button onClick={action('clicked')} variant="light" outline>
      light
    </Button>
    <Button onClick={action('clicked')} variant="dark" outline>
      dark
    </Button>
    <Button onClick={action('clicked')} variant="link" outline>
      link
    </Button>
  </>
);
export const Disable = () => (
  <>
    <Button onClick={action('clicked')} variant="primary" disabled>
      Primary
    </Button>
    <Button onClick={action('clicked')} variant="secondary" disabled>
      Secondary
    </Button>
    <Button onClick={action('clicked')} variant="success" disabled>
      Success
    </Button>
    <Button onClick={action('clicked')} variant="danger" disabled>
      danger
    </Button>
    <Button onClick={action('clicked')} variant="warning" disabled>
      warning
    </Button>
    <Button onClick={action('clicked')} variant="info" disabled>
      info
    </Button>
    <Button onClick={action('clicked')} variant="light" disabled>
      light
    </Button>
    <Button onClick={action('clicked')} variant="dark" disabled>
      dark
    </Button>
    <Button onClick={action('clicked')} variant="link" disabled>
      link
    </Button>
  </>
);
export const Size = () => (
  <>
    <Button onClick={action('clicked')} variant="primary">
      Default
    </Button>
    <Button onClick={action('clicked')} variant="primary" size="sm">
      Small
    </Button>
    <Button onClick={action('clicked')} variant="primary" size="lg">
      Large
    </Button>
  </>
);

const buttonOptions2 = [
  { id: '1', label: 'Cost ($)', value: 'Cost' },
  { id: '2', label: 'Retail ($)', value: 'Retail' },
];
export const TwoButtonsSwitch = () => <ButtonSwitch buttonOptions={buttonOptions2} onClick={action('value')} />;

const buttonOptions3 = [
  { id: '1', label: 'Cost ($)', value: 'Cost' },
  { id: '2', label: 'Retail ($)', value: 'Retail' },
  { id: '3', label: 'Retail (%)', value: 'Retail Percentage' },
];
export const ThreeButtonsSwitch = () => (
  <ButtonSwitch buttonOptions={buttonOptions3} onClick={action('value')} />
);
export const LargeButtonSwitch = () => (
  <ButtonSwitch size="lg" buttonOptions={buttonOptions3} onClick={action('value')} />
);

export const ButtonSwitchWithDefaultView = () => (
  <ButtonSwitch defaultView="Retail" buttonOptions={buttonOptions3} onClick={action('value')} />
);

const variant = ['primary', 'secondary', 'success', 'danger', 'warning', 'info', 'light', 'dark', 'link'];
const size = ['', 'lg', 'sm'];
export const usingKnobs = () => (
  <Button
    onClick={action('clicked')}
    variant={select('variant', variant, 'primary', 'GROUP-ID1')}
    outline={boolean('outline', false, 'GROUP-ID1')}
    disabled={boolean('disable', false, 'GROUP-ID1')}
    size={select('size', size, '', 'GROUP-ID1')}>
    Button
  </Button>
);
