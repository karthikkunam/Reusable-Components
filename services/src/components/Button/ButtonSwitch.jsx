import React, { useState } from 'react';
import PropTypes from 'prop-types';

const ButtonSwitch = props => {
  const [value, setValue] = useState(props.defaultView ? props.defaultView : props.buttonOptions[0].value);
  const { buttonOptions, onClick, className, size } = props;

  const onButtonSwitch = value => {
    setValue(value);
    onClick && onClick(value);
  };
  return (
    <div className={`component-example mb-3 ${className}`}>
      <div
        className={`btn-group btn-group-toggle ${size === 'lg' ? 'btn-group-lg' : 'btn-group-sm'}`}
        data-toggle="buttons">
        {buttonOptions &&
          buttonOptions.length > 0 &&
          buttonOptions.map((option, index) => {
            return (
              <label key={index} className={`btn btn-secondary ${option.value === value ? 'active' : ''}`}>
                <input
                  type="radio"
                  name="options"
                  id={option.id}
                  autoComplete="off"
                  checked={option.value === value}
                  onClick={() => {
                    onButtonSwitch(option.value);
                  }}
                />
                {option.label}
              </label>
            );
          })}
      </div>
    </div>
  );
};

ButtonSwitch.displayName = 'ButtonSwitch';
ButtonSwitch.defaultProps = {
  onClick: () => {},
  className: '',
};

ButtonSwitch.propTypes = {
  buttonOptions: PropTypes.arrayOf({
    id: PropTypes.string, 
    label: PropTypes.string, 
    value: PropTypes.string,
  }).isRequired,
  className: PropTypes.string,
  onClick: PropTypes.func,
  defaultView: PropTypes.string,
  size: PropTypes.oneOf(['lg']),
};

export default ButtonSwitch;
