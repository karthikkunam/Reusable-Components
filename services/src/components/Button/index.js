import Button from './Button';
import ButtonSwitch from './ButtonSwitch';

export {ButtonSwitch, Button};
