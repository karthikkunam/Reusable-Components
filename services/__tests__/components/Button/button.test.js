import React from 'react';
import Button from '../../../src/components/Button/Button.jsx';
import { shallow } from 'enzyme';

describe('Button Component', () => {
  it('renders correctly', () => {
      const props={
        variant: 'primary',
        disabled: false,
        outline: false,
        size: undefined,
        className: '',
        onClick: jest.fn(),
        children: undefined,
      };
    const component = shallow(<Button {...props}/>);
    expect(component).toMatchSnapshot();
  });
  it('renders correctly outline as true', () => {
    const props={
      variant: 'secondary',
      disabled: false,
      outline: true,
      size: undefined,
      className: '',
      onClick: jest.fn(),
      children: undefined,
    };
  const component = shallow(<Button {...props}/>);
  expect(component).toMatchSnapshot();
});
});
