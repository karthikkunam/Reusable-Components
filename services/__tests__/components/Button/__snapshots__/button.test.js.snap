// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders correctly 1`] = `
<button
  className="btn btn-primary"
  onClick={[MockFunction]}
/>
`;

exports[`Button Component renders correctly outline as true 1`] = `
<button
  className="btn btn-outline-secondary"
  onClick={[MockFunction]}
/>
`;
